File: Assets/Packages/core.debug/DebugWindow.lua
Vector2 = UnityEngine.Vector2
InputField = UnityEngine.UI.InputField
Text = UnityEngine.UI.Text
LayoutRebuilder = UnityEngine.UI.LayoutRebuilder
RectTransform = UnityEngine.RectTransform
Log = Ballance2.Utils.Log
LogLevel = Ballance2.Utils.LogLevel
CloneUtils = Ballance2.Sys.Utils.CloneUtils
UIAnchorPosUtils = Ballance2.Sys.UI.Utils.UIAnchorPosUtils
UIAnchor = Ballance2.Sys.UI.Utils.UIAnchor
UIPivot = Ballance2.Sys.UI.Utils.UIPivot
EventTriggerListener = Ballance2.Sys.UI.Utils.EventTriggerListener
GameStaticResourcesPool = Ballance2.Sys.Res.GameStaticResourcesPool
GameEventNames = Ballance2.Sys.Bridge.GameEventNames
GamePackage = Ballance2.Sys.Package.GamePackage

---@type GameLuaObjectHostClass
DebugWindow = {
  --Initvars
  CommandInputField = nil,---@type InputField
  FilterLogInputField = nil,---@type InputField
  LogContentView = nil,---@type RectTransform
  LogScrollView = nil,---@type ScrollRect
  LogStacktraceScrollView = nil,---@type ScrollRect
  LogStacktraceView = nil,---@type RectTransform
  LogStacktraceText = nil,---@type Text
  CheckBoxWarningAndErr = nil,---@type Toggle
  CheckBoxMessages = nil,---@type Toggle
  ToggleAutoScroll = nil,---@type ToggleEx

  logObserver = nil,
  logAutoScroll = true,
  logShowErrAndWarn = true,
  logShowMessage = true,
  logItems = {},
  logForceDisable = true,
  logFont = nil,
  logMaxCount = 256,
  logFilter = '',

  cmdServer = nil, 
}

function CreateClass_DebugWindow()
  function DebugWindow:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
  end

  ---@param thisGameObject GameObject
  function DebugWindow:Start(thisGameObject)
    self.logFont = GameStaticResourcesPool.FindStaticAssets('FontConsole')

    -- 注册日志观察者
    self.logObserver =
      Log.RegisterLogObserver(
      function(level, tag, message, stackTrace)
        if(self.logForceDisable) then
          return false
        end

        if(#self.logItems > self.logMaxCount) then
          local item = table.remove(self.logItems, 0)
          UnityEngine.Object.Destroy(item.go)
        end

        if #message >= 16384 then
          message = string.sub(message, 0, 16383)
        end
        if #message >= 32768 then
          message = string.sub(message, 0, 32767)
        end

        local logColor = GetLogColor(level)
        local t = table.concat({'<color=#', logColor, '>', Log.LogLevelToString(level), '/', tag, ' ', message, '</color>'})

        ---Add text
        local newEle = CloneUtils.CreateEmptyUIObjectWithParent(self.LogContentView)
        ---@type Text
        local text = newEle:AddComponent(UnityEngine.UI.Text)
        text.text = t
        text.font = self.logFont
        UIAnchorPosUtils.SetUIAnchor(newEle.transform, UIAnchor.Stretch, UIAnchor.Top)
        UIAnchorPosUtils.SetUIPivot(newEle.transform, UIPivot.TopLeft)

        ---日志点击事件
        ---@param go GameObject
        EventTriggerListener.Get(newEle).onClick = function (go)
          self:ShowLogStackTrace(go.transform:GetSiblingIndex() + 1)
        end

        table.insert(self.logItems, {
          go = newEle,
          message = message,
          level = level,
          stackTrace = stackTrace,
        })

        --重新布局
        self:RelayoutLogContent()

        --滚动到末尾
        if(self.logAutoScroll) then
          self.LogScrollView.normalizedPosition = Vector2(0, 0)
        end
      end,
      LogLevel.All)

    self.logForceDisable = false

    --命令服务的准备
    self.cmdServer = GameManager.Instance.GameDebugCommandServer

    --游戏退出事件
    GameManager.GameMediator:RegisterEventHandler(GamePackage.GetSystemPackage(), GameEventNames.EVENT_BEFORE_GAME_QUIT, 'DebugWindow', function ()
      self.logForceDisable = true
      Log.UnRegisterLogObserver(self.logObserver)
      return false
    end)

    --发送没有抓取到的日志
    Log.SendLogsInTemporary()
  end
  function DebugWindow:OnDestroy()
    self.logForceDisable = true
    Log.UnRegisterLogObserver(self.logObserver)
  end

  --[[
    工具函数
  ]]--

  ---布局日志界面
  ---@param self table
  function DebugWindow:RelayoutLogContent()
    local transform = self.LogContentView
    local c = transform.childCount
    local h = 0
    local logFilter = self.logFilter
    local logItems = self.logItems
    local logFilterEmpty = logFilter == ''
    local logShowErrAndWarn = self.logShowErrAndWarn
    local logShowMessage = self.logShowMessage
    for i = 0, c - 1, 1 do
      ---@type RectTransform
      local child = transform:GetChild(i)
      local log = logItems[i + 1]
      if(
        ((logShowErrAndWarn and (log.level == LogLevel.Warning or log.level == LogLevel.Error)) or
        (logShowMessage and (log.level == LogLevel.Info or log.level == LogLevel.Debug))) and  
        (logFilterEmpty or logFilter == string.find(log.message, logFilter))
      ) then
        child.gameObject:SetActive(true)
        child.anchoredPosition = Vector2(0, -h)
        UIAnchorPosUtils.SetUILeftBottom(child, 10, UIAnchorPosUtils.GetUIBottom(child))
        child.sizeDelta = Vector2(0, 15)
        h = h + 15
      else
        child.gameObject:SetActive(false)
      end
    end
    transform.sizeDelta = Vector2(transform.sizeDelta.x, h < 200 and 200 or h)
  end
  ---清空日志条目
  ---@param self table
  function DebugWindow:ClearLogContent()
    local transform = self.LogContentView
    local c = transform.childCount
    for i = c - 1, 0, -1 do
      UnityEngine.Object.Destroy(transform:GetChild(i).gameObject)
    end
    transform.sizeDelta = Vector2(transform.sizeDelta.x, 300)
  end
  ---显示日志StackTrace
  ---@param index number
  function DebugWindow:ShowLogStackTrace(index)
    local data = self.logItems[index]
    self.LogStacktraceText.text = table.concat({ data.message, '\n= StackTrace ====\n', data.stackTrace })
    LayoutRebuilder.ForceRebuildLayoutImmediate(self.LogStacktraceText.rectTransform);
    local size = self.LogStacktraceText.rectTransform.sizeDelta;
    self.LogStacktraceView.sizeDelta = Vector2(size.x + 40, size.y + 40)
    self.LogStacktraceScrollView.normalizedPosition = Vector2(0, 1)
  end
  function DebugWindow:SelectLastLog()
    self:ShowLogStackTrace(self.LogContentView.childCount)
  end
  ---AutoScroll更改
  function DebugWindow:OnAutoScrollChanged() 
    self.logAutoScroll = self.ToggleAutoScroll.isOn
  end
  ---显示错误Checkbox更改事件
  function DebugWindow:OnShowErrCheckChange()
    self.logShowErrAndWarn = self.CheckBoxWarningAndErr.isOn
    self:RelayoutLogContent()
  end
  ---显示信息Checkbox更改事件
  function DebugWindow:OnShowMessagesCheckChange()
    self.logShowMessage = self.CheckBoxMessages.isOn
    self:RelayoutLogContent()
  end
  ---筛选Input输入完成事件
  function DebugWindow:OnFilterInputEndInput()
    self.logFilter = self.FilterLogInputField.text
    self:RelayoutLogContent()
  end
  ---执行命令
  function DebugWindow:ExecCommand()
    if self.cmdServer:ExecuteCommand(self.CommandInputField.text) then
      self.CommandInputField.text = ''
      self:SelectLastLog()
    end
  end
  ---显示调试窗口
  function DebugWindow:ShowDebugOptWindow()
    GlobalDebugOptWindow:Show()
  end
  ---清空命令窗口
  function DebugWindow:ClearCommand()
    self.logWarnCount = 0
    self.logErrorCount = 0
    self.logInfoCount = 0
    --清空日志条目
    self:ClearLogContent()
    GlobalDebugToolbar.LuaSelf:Clear()
    --重新布局
    self:RelayoutLogContent()
  end

  return DebugWindow:new(nil)
end

File: Assets/System/Scripts/Services/DebugTools/DebugWindow.cs
using System.Collections.Generic;
using Ballance2.Res;
using Ballance2.Services;
using Ballance2.Services.InputManager;
using Ballance2.UI.Core.Controls;
using Ballance2.UI.Utils;
using Ballance2.Utils;
using UnityEngine;
using UnityEngine.UI;

/*
* Copyright(c) 2021  mengyu
*
* 模块名：     
* DebugTools.cs
* 
* 用途：
* 调试窗口控制类。
*
* 作者：
* mengyu
*/

namespace Ballance2
{
  public class DebugWindow : MonoBehaviour
  {
    public GameObject LogItemPrefab = null;
    public InputField CommandInputField = null;
    public InputField FilterLogInputField = null;
    public RectTransform LogContentView = null;
    public ScrollRect LogScrollView = null;
    public ScrollRect LogStacktraceScrollView = null;
    public RectTransform LogStacktraceView = null;
    public Text LogStacktraceText = null;
    public Toggle CheckBoxWarningAndErr = null;
    public Toggle CheckBoxMessages = null;
    public Toggle CheckBoxSysInfo = null;
    public Toggle CheckBoxMemInfo = null;
    public ToggleEx ToggleAutoScroll = null;

    private struct LogItem
    {
      public GameObject go;
      public string message;
      public LogLevel level;
      public string stackTrace;
    }
    private Font logFont;
    private int logObserver = 0;
    private bool logAutoScroll = true;
    private bool logShowErrAndWarn = true;
    private bool logShowMessage = true;
    private List<LogItem> logItems = new List<LogItem>();
    private bool logForceDisable = true;
    private int logMaxCount = 256;
    private string logFilter = "";

    private int commandHistoryIndex = 0;
    private List<string> commandHistory = new List<string>();
    private GameDebugCommandServer commandServer = null;
    private int commandInputListenKey1;
    private int commandInputListenKey2;
    private GameUIManager GameUIManager;

    private void Start()
    {
      this.GameUIManager = GameManager.Instance.GetSystemService<GameUIManager>();
      this.logFont = GameStaticResourcesPool.FindStaticAssets("FontConsole") as Font;

      // 注册日志观察者
      this.logObserver = Log.RegisterLogObserver((level, tag, message, stackTrace) =>
      {
        if (this.logForceDisable)
          return;

        if (this.logItems.Count > this.logMaxCount)
        {
          var nitem = logItems[0];
          this.logItems.RemoveAt(0);
          UnityEngine.Object.Destroy(nitem.go);
        }

        if (message.Length > 32768)
          message = message.Substring(0, 32767);

        var logColor = Log.GetLogColor(level);
        var t = string.Format("<color=#{0}>{1}/{2} {3}</color>", Log.GetLogColor(level), Log.LogLevelToString(level), tag, message);

        var newEle = CloneUtils.CloneNewObjectWithParent(LogItemPrefab, this.LogContentView);
        var text = newEle.GetComponent<Text>();
        text.text = t;
        UIAnchorPosUtils.SetUIAnchor((RectTransform)newEle.transform, UIAnchor.Stretch, UIAnchor.Top);
        UIAnchorPosUtils.SetUIPivot((RectTransform)newEle.transform, UIPivot.TopLeft);

        //-日志点击事件
        EventTriggerListener.Get(newEle).onClick = (go) => this.ShowLogStackTrace(go.transform.GetSiblingIndex());

        var item = new LogItem();
        item.go = newEle;
        item.message = message;
        item.level = level;
        item.stackTrace = stackTrace;
        this.logItems.Add(item);

        //重新布局
        this.RelayoutLogContent();

        text.gameObject.SetActive(true);
        //滚动到末尾
        if (this.logAutoScroll)
          this.LogScrollView.normalizedPosition = Vector2.zero;
      }, LogLevel.All);

      this.logForceDisable = false;

      //命令服务的准备
      this.commandServer = GameManager.Instance.GameDebugCommandServer;
      commandServer.RegisterCommand("clear", (keyword, fullCmd, argsCount, args) => {
        ClearCommand();
        commandHistory.Clear();
        commandHistoryIndex = 0;
        return true;
      }, 0, "clear 清空控制台");

      //发送没有抓取到的日志
      Log.SendLogsInTemporary();

      commandHistory.Add("");
      commandInputListenKey1 = this.GameUIManager.ListenKey(KeyCode.UpArrow, (KeyCode key, bool downed) => {
        if(downed && CommandInputField.isFocused) UpCommandHistory();
      });
      commandInputListenKey2 = this.GameUIManager.ListenKey(KeyCode.DownArrow, (KeyCode key, bool downed) => {
        if(downed && CommandInputField.isFocused) DownCommandHistory();
      });
    }
    private void OnDestroy()
    {
      if(this.GameUIManager != null) {
        this.GameUIManager.DeleteKeyListen(commandInputListenKey1);
        this.GameUIManager.DeleteKeyListen(commandInputListenKey2);
      }
      logForceDisable = true;
      Log.UnRegisterLogObserver(logObserver);
    }

    //布局日志界面
    private void RelayoutLogContent()
    {
      var transform = LogContentView;
      var h = 0;
      var logFilterEmpty = logFilter == "";
      for (var i = 0; i < transform.childCount; i++)
      {
        var child = transform.GetChild(i) as RectTransform;
        if(i >= logItems.Count)
          break;
        var log = logItems[i];
        if (
          (logShowErrAndWarn && (log.level == LogLevel.Warning || log.level == LogLevel.Error)) ||
          (logShowMessage && !(log.level == LogLevel.Warning || log.level == LogLevel.Error)) &&
          (logFilterEmpty || log.message.Contains(logFilter))
        )
        {
          child.gameObject.SetActive(true);
          child.anchoredPosition = new Vector2(0, -h);
          UIAnchorPosUtils.SetUILeftBottom(child, 10, UIAnchorPosUtils.GetUIBottom(child));
          child.sizeDelta = new Vector2(0, 20);
          h = h + 20;
        }
        else
          child.gameObject.SetActive(false);
      }
      transform.sizeDelta = new Vector2(transform.sizeDelta.x, h < 200 ? 200 : h);
    }
    //清空日志条目
    private void ClearLogContent()
    {
      var transform = LogContentView;
      var c = transform.childCount;
      for (var i = c - 1; i >= 0; i--)
        UnityEngine.Object.Destroy(transform.GetChild(i).gameObject);
      transform.sizeDelta = new Vector2(transform.sizeDelta.x, 300);
    }
    //显示日志StackTrace
    private void ShowLogStackTrace(int index)
    {
      var data = logItems[index];
      LogStacktraceText.text = data.message + "\n= StackTrace ====\n" + data.stackTrace;
      LayoutRebuilder.ForceRebuildLayoutImmediate(LogStacktraceText.rectTransform);
      var size = LogStacktraceText.rectTransform.sizeDelta;
      LogStacktraceView.sizeDelta = new Vector2(size.x + 40, size.y + 40);
      LogStacktraceScrollView.normalizedPosition = new Vector2(0, 1);
    }
    private void SelectLastLog()
    {
      ShowLogStackTrace(LogContentView.childCount - 1);
    }
    
    //AutoScroll更改
    public void OnAutoScrollChanged()
    {
      logAutoScroll = ToggleAutoScroll.isOn;
    }
    //显示错误Checkbox更改事件
    public void OnShowErrCheckChange()
    {
      logShowErrAndWarn = CheckBoxWarningAndErr.isOn;
      RelayoutLogContent();
    }
    //显示信息Checkbox更改事件
    public void OnShowMessagesCheckChange()
    {
      logShowMessage = CheckBoxMessages.isOn;
      RelayoutLogContent();
    }
    //筛选Input输入完成事件
    public void OnFilterInputEndInput()
    {
      logFilter = FilterLogInputField.text;
      RelayoutLogContent();
    }
    
    private void UpCommandHistory() {
      if(commandHistoryIndex == 0) {
        commandHistory[0] = CommandInputField.text;
        commandHistoryIndex = commandHistory.Count - 1;
      } else {
        if(commandHistoryIndex > 1) {
          commandHistory[commandHistoryIndex] = CommandInputField.text;
          commandHistoryIndex--;
          CommandInputField.text = commandHistory[commandHistoryIndex];
        }
      }
    }    
    private void DownCommandHistory() {
      if(commandHistoryIndex > 0) {
        if(commandHistoryIndex < commandHistory.Count - 1) {
          commandHistory[commandHistoryIndex] = CommandInputField.text;
          commandHistoryIndex ++;
          CommandInputField.text = commandHistory[commandHistoryIndex];
        } else {
          commandHistoryIndex = 0;
          CommandInputField.text = commandHistory[0];
        }
      }
    }

    public void ForceGC() {
      System.GC.Collect();
      Log.D("Debug", "GC Finish");
    }
    public void OnShowSysInfoCheckChange()
    {
      GameManager.Instance.GameActionStore.CallAction("DbgStatShowSystemInfo", CheckBoxSysInfo.isOn);
    }
    public void OnShowMemInfoCheckChange()
    {
      GameManager.Instance.GameActionStore.CallAction("DbgStatShowStats", CheckBoxMemInfo.isOn);
    }

    public void CopyStacktrace()
    {
      UnityEngine.GUIUtility.systemCopyBuffer = LogStacktraceText.text;
      GameUIManager.GlobalToast("已复制到剪贴板", 1);
    }
    //执行命令
    public void ExecCommand()
    {
      if (commandServer.ExecuteCommand(CommandInputField.text))
      {
        commandHistory.Add(CommandInputField.text);
        commandHistoryIndex = 0;
        CommandInputField.text = "";
        SelectLastLog();
      }
    }
    //清空命令窗口
    public void ClearCommand()
    {
      //清空日志条目
      ClearLogContent();
      GameManager.GameMediator.NotifySingleEvent("DebugToolsClear");
      //重新布局
      RelayoutLogContent();
    }
  }
}
